complex/ft_main.c

```diff
@@ -253,11 +254,100 @@ static int ft_fw_result_index(int fi_errno)
 	}
 }
 
-static int ft_fw_server(void)
+static int ft_fw_process_list(struct fi_info *hints, struct fi_info *info)
+{
+	int ret, subindex, result;
+	size_t len;
+
+	for (subindex = 1, fabric_info = info; fabric_info;
+	     fabric_info = fabric_info->next, subindex++) {
+
+		//check needed to skip utility providers, unless requested
+		if (!ft_util_name(hints->fabric_attr->prov_name, &len) &&
+		    strcmp(hints->fabric_attr->prov_name,
+		    fabric_info->fabric_attr->prov_name))
+			continue;
+
+		ret = ft_check_info(hints, fabric_info);
+		if (ret)
+			return ret;
+
+		ft_fw_update_info(&test_info, fabric_info, subindex);
+		printf("Starting test %d-%d: ", test_info.test_index,
+			subindex);
+		ft_show_test_info();
+
+		result = ft_run_test();
+		results[ft_fw_result_index(-result)]++;
+
+		ret = ft_sock_send(sock, &result, sizeof result);
+		if (result) {
+			FT_PRINTERR("ft_run_test", result);
+		}
+		else if (ret) {
+			FT_PRINTERR("ft_sock_send", ret);
+			return ret;
+		}
+	}
+	test_info.prov_name[0] = '\0';
+	ret = ft_sock_send(sock, &test_info, sizeof test_info);
+	if (ret) {
+		FT_PRINTERR("ft_sock_send", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ft_server_child()
 {
 	struct fi_info *hints, *info;
 	int ret;
 
+	hints = fi_allocinfo();
+	if (!hints)
+		return -FI_ENOMEM;
+
+	ft_fw_convert_info(hints, &test_info);
+	printf("Starting test %d:\n", test_info.test_index);
+
+	ret = fi_getinfo(FT_FIVERSION, ft_strptr(test_info.node),
+			 ft_strptr(test_info.service), FI_SOURCE,
+			 hints, &info);
+	if (ret && ret != -FI_ENODATA) {
+		FT_PRINTERR("fi_getinfo", ret);
+	} else {
+		/* fabric_info is replaced when connecting */
+		if (ret == -FI_ENODATA)
+			info = NULL;
+
+		ret = ft_fw_process_list(hints, info);
+
+		if (fabric_info != info)
+			fi_freeinfo(fabric_info);
+		fabric_info = NULL;
+	}
+	fi_freeinfo(info);
+
+	if (ret) {
+		FT_PRINTERR("ft_fw_process_list", ret);
+		printf("Node: %s\nService: %s\n",
+			test_info.node, test_info.service);
+		printf("%s\n", fi_tostr(hints, FI_TYPE_INFO));
+	}
```

Comment:
There's something off here.  fi_freeinfo() is called even if fi_getinfo() above fails, and the if (ret) check executes if fi_getinfo() fails too, resulting in a misleading print statement.  Maybe both statements belong inside the else?